### PR TITLE
[CP3.0] reimplement operandrequest mutating webhook

### PR DESCRIPTION
**Pull request by qpdpQ**
How to test:

Setting up dev cs-operator image on your cluster:

1. ensure you have future catalogsource on your cluster
2. export REGISTRY=quay.io/yuchen_li1
3. run `make build-dev-image`
4. edit manager.yaml `image: quay.io/yuchen_li1/common-service-operator-amd64:dev`
5. run `make build-bundle-image`
6. run `make run-bundle`

Testing:
1. search for service `ibm-common-service-webhook`
2. search for mutating webhook `ibm-common-service-webhook-configuration`
3. create an operandrequest, mutating webhook will set the correct namespace for it.

### CODE REVIEW

**Review by bitscuit — CHANGES_REQUESTED** [review]
Comment out this patch because it's empty, and it breaks `make bundle-manifests`

File: main.go [hunk 1]
```diff
@@ -213,13 +216,19 @@ func main() {
...
-	}).SetupWebhookWithManager(mgr); err != nil {
-		klog.Errorf("Unable to create OperandRequest webhook: %v", err)
-		os.Exit(1)
+
+	// Start up the webhook server
+	if err := setupWebhooks(mgr, bs); err != nil {
+		klog.Error(err, "Error setting up webhook server")
 	}
 
+	// if err = (&operandrequestwebhook.Defaulter{
```
[the remark this answers is not in the document]
Comment (qpdpQ) [reply main.go:225]: Sure!

**Comment by qpdpQ** [comment]
The `MutatingWebhookConfiguration` can be created now, but this `expected pointer, but got invalid kindError setting up webhook server` error still exists...
It happened on this function https://github.com/IBM/ibm-common-service-operator/pull/1023/files#diff-6242436e174eb22954722f8360a791a30d570d828b7ead1ec4708816ed34238cR139, still investigating on why it reports this error
I even can see the same error in the cs-webhook in CP2.0... and it works well for a long time, maybe it is at low risk.

BTW not sure if we need to keep this env var check in cs-operator: https://github.com/IBM/ibm-common-service-webhook/blob/master/cmd/manager/main.go#L143

**Comment by Daniel-Fan** [comment]
> BTW not sure if we need to keep this env var check in cs-operator: https://github.com/IBM/ibm-common-service-webhook/blob/master/cmd/manager/main.go#L143

I think we could keep this env variable available in cs-operator.
It is only for debugging purpose or we could easily disable the webhook deployment when something goes wrong.

If there is no env variable set, we should enable the webhook by default.

**Comment by Daniel-Fan** [comment]
> The MutatingWebhookConfiguration can be created now, but this `expected pointer, but got invalid kindError setting up webhook server` error still exists...

Trying to explore where this error comes from
- It returns from [`blder.Complete()`](https://github.com/kubernetes-sigs/controller-runtime/blob/1730628f118be1128787012c164379e6ac46c75f/pkg/builder/webhook.go#L57)
- `blder.Complete()` returns [`blder.registerWebhooks()`](https://github.com/kubernetes-sigs/controller-runtime/blob/1730628f118be1128787012c164379e6ac46c75f/pkg/builder/webhook.go#L71)
- `blder.registerWebhooks()` is possible to return an error from two functions
    - `blder.gvk, err = apiutil.GVKForObject(blder.apiType, blder.mgr.GetScheme())`
    - `err = blder.registerConversionWebhook()`
- Both functions have a call on [`scheme.ObjectKinds(obj)`](https://github.com/kubernetes-sigs/controller-runtime/blob/1730628f118be1128787012c164379e6ac46c75f/pkg/client/apiutil/apimachinery.go#L98) from `apimachinery` package
- Function [`ObjectKinds(obj Object)`](https://github.com/kubernetes/apimachinery/blob/a644435e2c133d990b624858d9c01985d7f59adf/pkg/runtime/scheme.go#L265) will return an error on [`v, err := conversion.EnforcePtr(obj)`](https://github.com/kubernetes/apimachinery/blob/a644435e2c133d990b624858d9c01985d7f59adf/pkg/runtime/scheme.go#L279)
- Function [`func EnforcePtr(obj interface{})`](https://github.com/kubernetes/apimachinery/blob/a644435e2c133d990b624858d9c01985d7f59adf/pkg/conversion/helper.go#L27) ensures that obj is a pointer of some sort. Returns a reflect.Value of the dereferenced pointer, ensuring that it is settable/addressable. Returns an error if this is not possible.
- And finally we will get an similar error msg `expected pointer, but got invalid kind`

Also there is a relative issue for DELETE operations on Mutating Webhook, not sure if it is what we could avoid. Currently I did not observe any deletion operation for OperandRequest mutating webhook
https://github.com/kubernetes/kubernetes/issues/66412

**Comment by Daniel-Fan** [comment]
> I even can see the same error in the cs-webhook in CP2.0... and it works well for a long time, maybe it is at low risk.

Personally, I think we could ignore this error for now, given it exists in previous webhook deployment and we did not hit any issues. Instead we could more focus on whether its functionality is working properly.

**Review by bitscuit — CHANGES_REQUESTED** [review]
Something the code does not consider right now is that we will need multiple webhookconfigurations, one per tenant.

Also, during testing I hit the following error

```
2023-02-16T19:34:14.162Z ERROR controller.commonservice Reconciler error {"reconciler group": "operator.ibm.com", "reconciler kind": "CommonService", "name": "common-service", "namespace": "ibm-common-services", "error": "unable to get: /ibm-operandrequest-webhook-configuration because of unknown namespace for the cache"}
```

This error does not happen when installed in AllNamespaces mode. I suspect this might be because when the operator is in OwnNamespaces mode its cache does not have cluster scoped resources in it. There might be some setting we have to explicitly set.

**Comment by Daniel-Fan** [comment]
```
2023-02-16T19:34:14.162Z ERROR controller.commonservice Reconciler error {"reconciler group": "operator.ibm.com", "reconciler kind": "CommonService", "name": "common-service", "namespace": "ibm-common-services", "error": "unable to get: /ibm-operandrequest-webhook-configuration because of unknown namespace for the cache"}
```

The error is due to the uncached webhook-configuration in cs's manager.
Add a customized Cache to store cluster-scoped webhook configurations.

**Comment by Daniel-Fan** [comment]
The test looks good so far.
But as Henry stated, we will need to support multiple webhook configurations in the same cluster, per tenant.

**Comment by qpdpQ** [comment]
change webhookconfiguration name unblock the multi-wehook test
two issues still in testing:
1. not sure if we need to change the webhook name as well 
https://github.com/IBM/ibm-common-service-operator/pull/1023/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R261
2. not sure if we should keep the rule of webhook as namespaceScoped 
https://github.com/IBM/ibm-common-service-operator/pull/1023/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R266
image

**Comment by Daniel-Fan** [comment]
> not sure if we need to change the webhook name as well
https://github.com/IBM/ibm-common-service-operator/pull/1023/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R261

I think we could keep this as is for now.

> not sure if we should keep the rule of webhook as namespaceScoped
https://github.com/IBM/ibm-common-service-operator/pull/1023/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R266

We could keep its `NamespaceScoped` given `OperandRequest` is a namespace-scoped resources

**Review by Daniel-Fan — APPROVED** [review]
/lgtm

**Pull request merged**